core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenome.java

```diff
@@ -0,0 +1,259 @@
+        genomeInternalIds.put(referenceGenome.getGenomeName(), referenceGenome.getReferenceGenomeId());
+    }
+
+    private static synchronized void reCache() {
+        clearCache();
+        Connection con = null;
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+        try {
+            con = JdbcUtil.getDbConnection(DaoReferenceGenome.class);
+            pstmt = con.prepareStatement("SELECT * FROM reference_genome");
+            rs = pstmt.executeQuery();
+            while (rs.next()) {
+                ReferenceGenome referenceGenome = extractReferenceGenome(rs);
+                addCache(referenceGenome);
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
+        } finally {
+            JdbcUtil.closeAll(DaoReferenceGenome.class, con, pstmt, rs);
+        }
+    }
+    
+    /**
+     * Add a new reference genome to the Database.
+     * @param referenceGenome   Reference Genome.
+     * @throws DaoException Database Error.
+     */
+    public static void addReferenceGenome(ReferenceGenome referenceGenome) throws DaoException {
+        reCache();
+        addReferenceGenome(referenceGenome, false);
+    }
+
+    /**
+     * Add a new reference genome to the Database.
+     * @param referenceGenome
+     * @param overwrite if true, overwrite if exist.
+     * @throws DaoException
+     */
+    public static void addReferenceGenome(ReferenceGenome referenceGenome, boolean overwrite) throws DaoException {
+
+        ReferenceGenome existing = getReferenceGenomeByInternalId(referenceGenome.getReferenceGenomeId());
+        if (existing!=null) {
+            if (!overwrite) {
+                throw new DaoException("Reference Genome " + referenceGenome.getBuildName() + "is already imported.");
+            } else {
+                updateReferenceGenome(referenceGenome);
+            }
+        }
+        else {
+            Connection con = null;
+            PreparedStatement pstmt = null;
+            ResultSet rs = null;
+            try {
+                con = JdbcUtil.getDbConnection(DaoReferenceGenome.class);
+                pstmt = con.prepareStatement("INSERT INTO reference_genome " +
+                        "( `species`, `name`, "
+                        + "`build_name`, `genome_size`, `URL`, "
+                        + " `release_date` ) VALUES (?,?,?,?,?,?)",
+                    Statement.RETURN_GENERATED_KEYS);
+                pstmt.setString(1, referenceGenome.getSpecies());
+                pstmt.setString(2, referenceGenome.getGenomeName());
+                pstmt.setString(3, referenceGenome.getBuildName());
+                pstmt.setLong(4, referenceGenome.getGenomeSize());
+                pstmt.setString(5, referenceGenome.getUrl());
+                pstmt.setDate(6, new java.sql.Date(referenceGenome.getReleaseDate().getTime()));
+                if (pstmt.executeUpdate() != -1 ) {
```

Comment:
the return value from executeUpdate is never -1 .. it will be 0 if no records were altered, or it will be a record count if some records were altered. So I think this statement should be 
if (pstmt.executeUpdate() != 1 )
Also, please add an "else" clause to this statement which something like:
else {
throw new DaoException("attempt to add new referenceGenome record failed")
}

Reply:
good catch! Will make changes.

Reply:
changed. Progress monitor indicates update statement that affected zero rows.
no need for an else statement since try..catch block will handle any exceptions.
```
            if (pstmt.executeUpdate() == 0) {
                ProgressMonitor.logWarning("No change for " + gene.getEntrezGeneId() + " " + gene.getReferenceGenomeId());
            }
```